### PR TITLE
Input updates

### PR DESCRIPTION
**Pull request by mrbobbybobberson**
This is a set of updates (primarily minor bug fixes and maintainability refactorings) we took on our recent project.  I'd like to contribute them if others find them valuable.  Thanks!

### CODE REVIEW

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs [hunk 1]
```diff
@@ -54,12 +54,10 @@ public enum CursorStateEnum
         public CursorStateEnum CursorState { get { return cursorState; } }
         private CursorStateEnum cursorState = CursorStateEnum.None;
 
-#pragma warning disable 0649
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Rather than disabling the warning, we made the field "protected" which has an added benefit that if a child class introduces a new SerializedField with the same name, the compiler will help us catch it, rather than leaving Unity to silently ignore the issue and leave us developers confused.

Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: What was this warning we were suppressing?

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 2]
```diff
@@ -156,6 +167,10 @@ public void RegisterPointer(IPointingSource pointingSource)
                 {
                     gazeManagerPointingData = new PointerData(pointingSource);
                 }
+                else
+                {
+                    gazeManagerPointingData.ResetFocusedObjects();
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: This fixes a bug.  When switching between a 6DoF controller's pointing ray and gaze, if the user is staring at an object, the focus wasn't updated correctly.  This change fixes that case.

File: Assets/HoloToolkit/Input/Scripts/InputEventData/PointerInputEventData.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/InputEventData/PointerInputEventData.cs]: This fixes a bug.  When UnityUI events were flowing, they weren't correctly "pointer-stealing" between 6DoF pointing ray and gaze.  By attaching InputSource and SourceId to those events, too, "pointer-stealing" works correctly.

File: Assets/HoloToolkit/Utilities/Scripts/Extensions/Extensions.cs [hunk 4]
```diff
@@ -1,6 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
+using System;
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Utilities/Scripts/Extensions/Extensions.cs]: None of these changes are necessary for input to work, but we found them useful while developing our app.  I'm fine keeping them or removing them.  Anyone feel strongly either way?

File: Assets/HoloToolkit/Utilities/Scripts/HashCodes.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Utilities/Scripts/HashCodes.cs]: HashCodes isn't necessary for input to work, but we found them useful while developing our app. I'm fine keeping them or removing them. Anyone feel strongly either way?

File: Assets/HoloToolkit/Utilities/Scripts/ReadOnlyHashSet.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Utilities/Scripts/ReadOnlyHashSet.cs]: ReadOnlyHashSet isn't necessary for input to work, but we found it useful while developing our app. I'm fine keeping it or removing it. Anyone feel strongly either way?

File: Assets/HoloToolkit/Utilities/Scripts/StartAwareSingleton.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Utilities/Scripts/StartAwareSingleton.cs]: StartAwareSingleton isn't necessary for input to work, but it is analogous to StartAwareBehaviour which is already in HoloToolkit, so I feel we should take this addition.  It was very useful when making our app.  Anyone feel strongly that we shouldn't?

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs [hunk 8]
```diff
@@ -230,10 +228,7 @@ protected virtual void RegisterManagers()
             InputManager.Instance.InputEnabled += OnInputEnabled;
             InputManager.Instance.InputDisabled += OnInputDisabled;
 
-            if (FocusManager.IsInitialized)
```
Comment (keveleigh) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Is this check not needed anymore?

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs [hunk 9]
```diff
@@ -278,16 +273,10 @@ private void TryLoadPointerIfNeeded()
                 // pointer currently registered with FocusManager, we use it.
 
                 Pointer = FocusManager.Instance.TryGetSinglePointer();
-
-                if (Pointer == null)
```
Comment (keveleigh) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Lines 281 to 290 were added during Build to fix/workaround some bugs. @NeerajW should have more insight.

File: Assets/HoloToolkit/Input/Scripts/InputSources/IInputSource.cs [hunk 10]
```diff
@@ -8,19 +8,6 @@
 namespace HoloToolkit.Unity.InputModule
 {
     /// <summary>
-    /// Flags used to indicate which input events are supported by an input source.
```
Comment (keveleigh) [inline Assets/HoloToolkit/Input/Scripts/InputSources/IInputSource.cs]: Are these removed intentionally?

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 8 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Correct.  FocusManager is a required component just as much as InputManager is, so if we somehow end up with a scene without a FocusManager in it, we want a big loud error just like we get when InputManager is missing so we can fix it.

File: Assets/HoloToolkit/Input/Scripts/InputSources/IInputSource.cs (hunk 10 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/InputSources/IInputSource.cs]: Yeah, the diffs line up kind of funny here, but this whole set of flags was made obsolete in an earlier refactoring of InputManager that someone did.  We removed the flags, since they weren't used anymore.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 9 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Let me know what the verdict is on this, @NeerajW.  Ideally all our prefabs and test scenes have FocusManager setup correctly in them already (I know a couple months back, I made sure they all did in our project, but perhaps those changes didn't get propagated back into all tests scenes in this repo.  Let me know if you want me to check and fix them all.

File: Assets/HoloToolkit/Input/Tests/Scenes/OverrideFocusedObjectTest.unity [hunk 11]
```diff
@@ -1,478 +0,0 @@
-%YAML 1.1
-%TAG !u! tag:unity3d.com,2011:
```
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Tests/Scenes/OverrideFocusedObjectTest.unity]: Why is this deleted? This functionality is used by several academy apps.

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs (hunk 2 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: I was just about to add a message about that :).  I think that should be the default behavior.  Are you OK with me making the main InputManagerTest scene 6DoF-compatible and make the main InputManager prefab include the SimpleSinglePointerSelector and be configured correctly for 6DoF or GGV?

File: Assets/HoloToolkit/Input/Tests/Scenes/OverrideFocusedObjectTest.unity (hunk 11 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Tests/Scenes/OverrideFocusedObjectTest.unity]: Ahh, perhaps I misunderstood.  I was under the impression that the OverrideFocusedObject was introduced as a temporary workaround for a pointer selection bug in the past.  My memory is just hazy though.  Did I remember that wrong?

Using a global listener or a modal input handler seems like the right way to handle override scenarios, was there some scenario they couldn't satisfy?

File: Assets/HoloToolkit/Input/Tests/Scripts/TestOverrideFocusedObject.cs [hunk 12]
```diff
@@ -1,31 +0,0 @@
-﻿using UnityEngine;
-
-namespace HoloToolkit.Unity.InputModule.Tests
-{
-    public class TestOverrideFocusedObject : MonoBehaviour, IInputClickHandler
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Tests/Scripts/TestOverrideFocusedObject.cs]: <This thread continued in OverrideFocusedObjectTest.unity>

File: Assets/HoloToolkit/Input/Scripts/InputManager.cs [hunk 13]
```diff
@@ -16,13 +16,6 @@ namespace HoloToolkit.Unity.InputModule
     /// </summary>
     public class InputManager : Singleton<InputManager>
     {
-        /// <summary>
-        /// To tap on a hologram even when not focused on,
-        /// set OverrideFocusedObject to desired game object.
-        /// If it's null, then focused object will be used.
-        /// </summary>
-        public GameObject OverrideFocusedObject { get; set; }
```
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/InputManager.cs]: (This thread continued in OverrideFocusedObjectTest.unity)

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 9 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Will do.  It looks like the InputManager prefab wasn't updated to use SimpleSinglePointerSelector, so I figure that could be one of the causes, but I'll dig in.

**Comment by StephenHodgson** [comment]
@mrbobbybobberson, how will these changes relate to the master branch?

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs (hunk 2 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: I did this as part of the latest commits.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 9 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Yeah, by adding SimpleSinglePointerSelector to InputManager, that fixed the issues.  There were a few scenes that didn't have a cursor, but still used InputManager, so for those scenes, I disabled SimpleSinglePointerSelector accordingly.

File: Assets/HoloToolkit/Input/Scripts/InputManager.cs (hunk 13 above)
Comment (mrbobbybobberson) [inline Assets/HoloToolkit/Input/Scripts/InputManager.cs]: I did this as part of the latest commits.  Thanks for the catch!

**Comment by mrbobbybobberson** [comment]
@StephenHodgson, that question is probably best answered by @NeerajW .  I believe at some point the branches will merge, but I don't know when that will be.

**Comment by StephenHodgson** [comment]
Yup we've already got an issue for that: https://github.com/Microsoft/HoloToolkit-Unity/issues/707

I just don't want one giant merge at once.  It'll be best to do a bit at a time.
